cpmpy/expressions/core.py

```diff
@@ -744,3 +744,196 @@ def _wsum_make(arg):
         return [-1], [arg.args[0]]
     # default
     return [1], [arg]
+
+
+# Custom dictionary and set classes for CPMpy expressions.
+# These do not inherit from python dict or set, because python abstracts from the underlying hash table.
+# Meaning we cannot access it ourselves.
+# This makes you rely on super.add, which is what we are trying to avoid..
+class cpm_dict:
+    def __init__(self):
+        # Maps hash values to lists of (key, value) pairs with that hash
+        self._hash_groups = {}
+
+    def same(self, key1, key2):
+        """
+        Custom key comparison logic (override this as needed).
+        Example: Compare keys as strings to allow "1" and 1 to coexist.
+        """
+        return str(key1) == str(key2)
+
+    def _get_hash_group(self, key):
+        """Returns the list of (key, value) pairs for a given hash."""
+        h = hash(key)
+        if h not in self._hash_groups:
+            self._hash_groups[h] = []
+        return self._hash_groups[h]
+
+    def __setitem__(self, key, value):
+        hash_group = self._get_hash_group(key)
+        # Check for existing key using `same`
+        for i, (existing_key, _) in enumerate(hash_group):
+            if self.same(existing_key, key):
+                hash_group[i] = (key, value)  # Update existing entry
+                return
+        # Add new key-value pair if no collision
+        hash_group.append((key, value))
+
+    def __getitem__(self, key):
+        hash_group = self._get_hash_group(key)
```

Comment:
its a bit strange that this becomes a mutable operation: it will create an empty bucket for this hash if the hash has not been stored yet...

```
  h = hash(key)
  if h in self._hash_groups:
    for existing_key, value in self._hash_groups:
       if self.same(existing_key, key):
          return value
   raise...
```
I don't know, I'm not sure you really need the _get_hash_group() in any of the functions, but if you think its cleaner it wont matter too much I guess